feat(aggregation_mode): aggregate proofs in chunks

MarcosNicolau opened the pull request:
## Description

Split the aggregation of proofs in chunks. For this, the aggregation program had to be split in two:
1. User proofs: this runs for each one of the chunks (each chunk is configured to have a max size of 512)
    - Aggregates proofs.
    - Commits the Merkle root of each chunk as a public input.
2. Chunk aggregator: 
    - Takes all the chunks proof along with their corresponding leaves (proofs commitments).
    - Verifies that each proof originated from the chunk aggregator program.
    - Verifies that the merkle root from the proof public input matches the merkle root from the provided proof leaves.
    - Computes the Merkle root of all leaves and commits it as a public input.
    - Produces the final aggregated proof that is verified on-chain.

Note: Because the root aggregator needs to verify that all input proofs were produced by the chunk aggregator, it must know the chunk aggregator’s program ID. This introduces a dependency between the two programs, requiring the root aggregator to be built after the chunk aggregator.

A script was created to automates this process. Running make `agg_mode_write_program_ids` now will:
1. Compile programs.
2. Copy the chunk aggregator's program ID and replace it in the root aggregator program as a constant.
3. Recompile programs (this step is necessary since changing the root aggregator's constant affects its own program ID).

**Considerations**
- `Hash32` type and `MerkleTreeBackend` is defined three times:
    - In the sdk
    - In sp1 root aggregator program
    - In Risc0 root aggregator program
  We should consider having a `types` crates to share this kind of types. We can tackle this once we re-define the structure for the rust projects to have a single workspace with a `crates` folder containing them all. 

**How to test**

1. Start a local devnet:

```shell
make ethereum_package_start
```

2. Run batcher
```
make batcher_start_ethereum_package
```

3. Send many proofs to the batcher:

```shell
# Run it as much as you want to process more chunks
make batcher_send_sp1_burst 
# Leave this in the background to generate batches with different merkle roots 
batcher_send_burst_groth16
```

4. Wait some time until enough proofs have been sent to run the proof aggregator with many chunks (consider lowering the `MAX_PROOFS_PER_AGGREGATION` variable to wait less):

```shell
make start_proof_aggregator_gpu_ethereum_package AGGREGATOR=sp1
make start_proof_aggregator_gpu_ethereum_package AGGREGATOR=risc0
```

5. You should see via the logs the number of chunks to be processed. You can also inspect the gpu usage via `nvtop`.

6. Finally, once the aggregator finishes, you can verify that one of the proofs you sent has been aggregated via the aligned-cli and sdk:
```
cd batcher/aligned
# For SP1
cargo run verify-agg-proof \
		--network devnet \
		--from-block 0 \
		--proving_system SP1 \
		--public_input ../../scripts/test_files/sp1/sp1_fibonacci_4_1_3.pub \
		--program-id-file ../../scripts/test_files/sp1/sp1_fibonacci_4_1_3.vk \
		--beacon_url http://localhost:58801 \
		--rpc_url http://localhost:8545

# For Risc0
cargo run verify-agg-proof \
		--network devnet \
		--from-block 0 \
		--proving_system Risc0 \
		--program-id-file ../../scripts/test_files/risc_zero/fibonacci_proof_generator/fibonacci_id_2_0.bin \
		--public_input ../../scripts/test_files/risc_zero/fibonacci_proof_generator/risc_zero_fibonacci_2_0.pub \
		--beacon_url http://localhost:58801 \
		--rpc_url http://localhost:8545
```
## Type of change

- [x] New feature

## Checklist

- [ ] “Hotfix” to `testnet`, everything else to `staging`
- [ ] Linked to Github Issue
- [ ] This change depends on code or research by an external entity
  - [ ] Acknowledgements were updated to give credit
- [ ] Unit tests added
- [ ] This change requires new documentation.
  - [ ] Documentation has been added/updated.
- [ ] This change is an Optimization
  - [ ] Benchmarks added/run
- [ ] Has a known issue
  - [Link to the open issue addressing it]() 
- [ ] If your PR changes the Operator compatibility (Ex: Upgrade prover versions)
  - [ ] This PR adds compatibility for operator for both versions and do not change batcher/docs/examples
  - [ ] This PR updates batcher and docs/examples to the newer version. This requires the operator are already updated to be compatible

MarcosNicolau committed:
feat: risc0 aggregation program

MarcosNicolau committed:
feat: [wip] risc0 aggregator backend

MarcosNicolau committed:
feat: ProofAggregationService risc0 verification

MarcosNicolau committed:
feat: error handling + fetch risc0 proofs

MarcosNicolau committed:
feat: send risc0 aggregated proof to AlignedProofAggregationService contract

MarcosNicolau committed:
feat: deploy AlignedProofAggregationService contract with risc0 verifier

MarcosNicolau committed:
feat: allow risc0 succinct proofs

MarcosNicolau committed:
chore: modified risc0 no_pub_inputs to generate a succinct proof instead of a composite

MarcosNicolau committed:
feat: fetch proofs based on zkvm engine

MarcosNicolau committed:
chore: commands to start proof agregator based on verifier

MarcosNicolau committed:
fix: compute image id bytes in little endian not in be

MarcosNicolau committed:
feat: local verification of aggregated risc0 proof

MarcosNicolau committed:
Merge branch 'staging' into feat/aggregation-mode-risc0

MarcosNicolau committed:
chore: add gpu feature flag to run with cuda

MarcosNicolau committed:
docs: add gpu command

MarcosNicolau committed:
chore: undo no_pub_input file changes

MarcosNicolau committed:
ci: fix contracts build + install risc0 toolchain

MarcosNicolau committed:
chore: better proof aggregator makefile targets

MarcosNicolau committed:
docs: update how to run proof aggregator readme

MarcosNicolau committed:
ci: skip build on clippy

MarcosNicolau committed:
chore: address clippy warnings

MarcosNicolau committed:
fix: correctly setting risc0 tiny-keccak patch

MarcosNicolau committed:
chore: rename prove feature to proving to prevent collision with risc0 features

MarcosNicolau committed:
Revert "chore: rename prove feature to proving to prevent collision with risc0 features"

This reverts commit bd5ad9df9ed97cac7b40836c23d6d966c3b916e7.

MarcosNicolau committed:
chore: address juli's comments

MarcosNicolau committed:
refactor: replace sp1,risc0 features for .env variables

MarcosNicolau committed:
chore: update makefile for new aggregator env config

MarcosNicolau committed:
refactor: aggregation mode reduce amount of types and abstractions

MarcosNicolau committed:
Merge branch 'staging' into refactor/agg-mode-redundancies

MarcosNicolau committed:
feat: aggregate proofs in chunks of 512 proofs

MarcosNicolau committed:
fix: aggregated proofs commitment for the second layer

MarcosNicolau committed:
chore: add comments for aggregation function

MarcosNicolau committed:
chore: add tracing for chunk aggregation

MarcosNicolau committed:
Merge branch 'staging' into feat/aggregate-proofs-in-chunks

MauroToscano reviewed (changes requested):
We need to double check the soundness of this

MauroToscano commented:
This is unsound, you need to check if the aggregated proof came from the right ELF, you can test hacking it with another program that has the right public inputs but it's not fro man Agg proof. Also, don't do something that commits and with an if you don't commit in one case, make it more explicit

MarcosNicolau committed:
Merge branch 'staging' into feat/aggregate-proofs-in-chunks

MarcosNicolau committed:
feat: split aggregation in two programs

MauroToscano committed:
Merge and fix sp1

MarcosNicolau committed:
feat: update risc0 programs to correctly compute the merkle tree

MarcosNicolau committed:
feat: update sp1 programs to correctly compute the merkle tree

MarcosNicolau committed:
feat: aggregator backends for new programs

MarcosNicolau committed:
feat: update write_program_image_id_vk_hash with new programs

MarcosNicolau committed:
chore: build program and update chunk aggregator program id in root aggregators

MarcosNicolau committed:
Merge remote-tracking branch 'origin/staging' into feat/aggregate-proofs-in-chunks

MarcosNicolau committed:
chore: update anvil state

MarcosNicolau committed:
chore: address warnings and increase max proofs per aggregation to 512

MarcosNicolau committed:
Merge branch 'staging' into feat/aggregate-proofs-in-chunks

MarcosNicolau committed:
chore: script to build programs and write the proper program ids

MarcosNicolau committed:
chore: address clippy warnings

MarcosNicolau committed:
docs: aggregation mode how to compile instructions

MarcosNicolau committed:
chore: add docs for merkle tree backend for Hash32 in root programs

MarcosNicolau committed:
chore: update docs in functions

MarcosNicolau committed:
chore: make sp1 aggregation work on dev mode

MarcosNicolau committed:
chore: agg_mode_write_program_ids run build_programs script

MarcosNicolau committed:
fix: compilation issues on cuda

MarcosNicolau committed:
Merge branch 'staging' into feat/aggregate-proofs-in-chunks

MarcosNicolau committed:
chore: update ethereum package and anvil state

MarcosNicolau committed:
chore: rename program names

ex proof_aggregator now is user_proots_aggregator and ex root_aggregator is now chunk_aggregator

MarcosNicolau committed:
chore: deploy aligned conctrats with new program ids

MarcosNicolau committed:
chore: fix sp1 USER_PROOFS_AGGREGATOR_PROGRAM_VK_HASH constant

MarcosNicolau committed:
chore: redeploy anvil contracts for new program ids

MarcosNicolau committed:
chore: fix risc0_user_proofs_aggregator_image_id

MarcosNicolau committed:
chore: update ethereum package contracts

MarcosNicolau committed:
chore: logs typos

MarcosNicolau committed:
fix: network params ethereum package

JuArce commented on aggregation_mode/src/aggregators/mod.rs:
```diff
@@ -11,6 +11,9 @@ use sha3::{Digest, Keccak256};
 };
+use tracing::info;
+
+const MAX_PROOFS_PER_AGGREGATION: usize = 512;
```
It could be a config parameter, but we can modify this in another PR

MarcosNicolau replied:
Solved in [46ebd22](https://github.com/yetanotherco/aligned_layer/pull/1896/commits/46ebd22f42c5a49fc3e332082b1ddc4c2ebd1196).

JuArce commented on aggregation_mode/aggregation_programs/sp1/src/chunk_aggregator_main.rs:
```diff
@@ -0,0 +1,50 @@
+        let vkey = proof.vk;
+        let public_values_digest = Sha256::digest(&proof.public_inputs);
+
+        // Ensure the aggregated chunk originates from the L1 aggregation program.
```
```suggestion
        // Ensure the aggregated chunk originates from the user proofs aggregation program.
```

JuArce commented on contracts/script/deploy/config/devnet/proof-aggregator-service.devnet.config.json:
```diff
@@ -1,13 +1,13 @@
   "address": {
     "sp1VerifierAddress": "0x06530DD13aFB697B81E2d602a279EBAEC8Ca2E31",
-    "risc0VerifierAddress": "0x14C66A76b06A414e62048c50e4d21B295ad2Cb0C",
+    "risc0VerifierAddress": "0x663F3ad617193148711d28f5334eE4Ed07016602",
```
Why does this address change?

MarcosNicolau replied:
Honestly, I don't really know. But `Risc0` verification was reverting until I realize it was the address of the verifier.

MarcosNicolau committed:
Update aggregation_mode/aggregation_programs/sp1/src/chunk_aggregator_main.rs

Co-authored-by: Julian Arce <52429267+JuArce@users.noreply.github.com>

MarcosNicolau committed:
feat: proofs per chunk as a config param instead of const

JuArce committed:
nit: add missing new lines

MarcosNicolau committed:
chore: update comments
